Add some examples and the beginnings of the grammar.

Jackojc opened the pull request.

Jackojc committed:
Add some examples and the beginnings of the grammar.

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+unicode_escape = '\u' hex+
+hex_escape = '\x' hex+
+bin_escape = '\b' ( '0' | '1' )+
+escapes = '\n' | '\t' | '\r' | '\v' | '\f' | '\b' | '\'' | '\"' | '\\' | unicode_escape | hex_escape | bin_escape
```
Looks like the escaped single quote isnt recognized as being escaped...Maybe use `"'"` instead of `'\''`

noxabellus replied:
Looks like the escaped single quote isnt recognized as being escaped...Maybe use `"'"` instead of `'\''`

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+bin_literal = '0b' ( '0' | '1' )+
+integer_literal = ( digit | '_' )+
+float_literal = ( digit | '_' ) [ '.' ( digit | '_' )* ] 'f';
+string_literal = '"' ( escapes | utf8 )* '"'
```
Shouldnt string literals have the same optional prefixing `[ "utf8" | "utf16" | "utf32" ]` as the char literals?

Jackojc replied:
Yeah, whoops

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+char_literal = [ "utf8" | "utf16" | "utf32" ] "'" utf8 "'"
+
+number_literal = integer_literal | float_literal | hex_literal | bin_literal
+literal = number_literal | string_literal | unicode_literal | char_literal
```
What is `unicode_literal`?

Jackojc replied:
Oh whoops, remnant from an older iteration

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+
+
+# Statements
+declaration = identifier ( ',' identifier )* ':' [ type ] [ '=' expr ]
```
This makes `my_var:;` a valid declaration

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+
+# Statements
+declaration = identifier ( ',' identifier )* ':' [ type ] [ '=' expr ]
+assignment = identifier '=' ( identifier | literal )
```
This doesnt allow assigning to dereferenced pointers, subscripts etc; and it doesnt allow assigning arbitrary expressions, such as the result of a function call. Think this should be `expr '=' expr`. This isnt precisely accurate as not all expressions are valid on the left side, but theres such a large amount as to be impractical to address (consider `@fn_call()[sub] = ...` for example ... Validating the left side is more of a semantic problem than a grammatical one

dekrain replied:
so
```ebnf
assignment = (identifier | '@' expr) '=' ( identifier | literal )
```

noxabellus replied:
Nah, that doesnt cover subscript, field access, etc. Like I said, theres too many possibilities to properly address in the grammar. Its a semantic problem.

Jackojc replied:
I don't think that solves array subscripts dekrain, maybe it is just better to use `expr '=' expr` and check if its correct later on

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+sum = 'sum' identifier '{' sum_entry ( ',' sum_entry )* [ ',' ] '}'
+
+enum_entry = identifier [ '=' literal ]
+enum = 'enum' '{' enum_entry ( ',' enum_entry )* [ ',' ] '}'
```
Missing name for enum

Jackojc replied:
What do you mean?

noxabellus replied:
I would assume enum variants would be scoped rather than poluting the namespace like C. In other words enums would be defined like
```
enum MyEnum { A, B, C }
```
and referenced like `MyEnum.A`

dekrain replied:
wouldn't we do
```
type MyEnum : enum { A, B, C }
```

noxabellus replied:
I think what we discussed for `type` aliases was that the right hand side of the `:` is a type expression, and enum isnt a valid type expression

Jackojc replied:
I think it's best to keep it as to ensure that the enum is properly namespaced, if we make it a type expression, you would have an enum which pollutes global scope.

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+
+
+# Expressions
+type_compare = identifier 'is' ( [ '.' ] type )
```
This should be `variant_compare` and the thing after the `.` isn't a type, its just an identifier

Jackojc replied:
`type` can be an identifier also, we may want to check something like `foo is i8` etc. Will change the name though.

noxabellus replied:
> `type` can be an identifier also, we may want to check something like `foo is i8` etc. Will change the name though.

Fully disagree with this. In the current version plan (without generics) `foo is i8` is entirely pointless, and even if we wanted that later, it still would need to be separate cases `('.' identifier) | type`

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+
+# Expressions
+type_compare = identifier 'is' ( [ '.' ] type )
+cast = identifier 'as' ( [ '.' ] type )
```
Don't need the `.` here in the case of a type cast, only for variant casts. Should be something like `(('.' identifier) | type)`

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+type_compare = identifier 'is' ( [ '.' ] type )
+cast = identifier 'as' ( [ '.' ] type )
+
+call = identifier '(' expr ( ',' expr )* ')' ';'
```
Need to allow calling other kinds of expressions, not sure what this should look like. Probably just expr instead of identifier?

Jackojc replied:
Oh I suppose this would be useful if we add higher-order functions and want to support things like `foo()()`

noxabellus replied:
Yeah also arrays of functions, etc

dekrain commented on examples/ack.mlk:
```diff
@@ -0,0 +1,23 @@
+# Ackermann function (http://rosettacode.org/wiki/Ackermann_function)
+
+fn ackermann(m, n: int) int {
+	if (m == 0) {
+		n + 1
+	}
+
+	if (n == 0) {
+		ackermann(m - 1, 1)
+	}
+
+	ackermann(m - 1, ackermann(m, n - 1))
+}
```
Fixed version
```
fn ackermann(m, n: int) int {
    if m == 0 {
        n + 1
    } else if n == 0 {
        ackermann(m - 1, 1)
    } else {
        ackermann(m - 1, ackermann(m, n - 1))
    }
}
```

noxabellus replied:
Unnecessary parens in the predicate expressions and invalid use of trailing expressions... Need some `else`s

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+if = 'if' expr block [ 'else' block ]
+
+match = 'match' expr '{' match_case* '}'
+match_case = ( ( '.' identifier [ '@' identifier ] ) | ( identifier | literal ) | '*' ) block
```
Expression match arms should allow any arbitrary constant expression. Whether its constant or not is a semantic problem. So this should just be `... | expr | '*' )`

noxabellus replied:
We probably should support multiple, comma separated case predicates per block
```ebnf
match = 'match' expr '{' match_case* '}'
match_predicate = ( ( '.' identifier [ '@' identifier ] ) | expr | '*' )
match_case =  match_predicate ( ',' match_predicate )* block
```

dekrain commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+
+function = 'fn' identifier '(' declaration ( ',' declaration )* ')' [ type ] block
+
+for = 'for' identifier [':' type] '=' expr ',' expr [',' expr] block
```
I think, supported by the examples, that it should be
```ebnf
for = 'for' identifier ':' [type] '=' expr ',' expr [',' expr] block
```

Jackojc replied:
Good catch

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+
+
+# Misc
+expr = match | if | call | cast | type_compare | math
+statement = ( for | while ) | ( ( declaration | expr | assignment ) ';' )
```
Shouldnt these just be under their respective categories?

noxabellus replied:
Missing:
return
break, continue
goto, label

dekrain commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+# Syntactical Grammar
+multiline_comment = '#[' ( utf8 | multiline_comment )* ']'
+
+unicode_escape = '\u' hex+
+hex_escape = '\x' hex+
```
I think it better works as
```ebnf
unicode_escape = '\u' hex{4} | '\U' hex{8}
hex_escape = '\x' hex{2}
```

Jackojc replied:
Yeah I wasn't sure what notation to use for it but I guess that works.

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+# Misc
+expr = match | if | call | cast | type_compare | math
+statement = ( for | while ) | ( ( declaration | expr | assignment ) ';' )
+block = '{' statement* [ expr ] '}'
```
Block is a valid expression with trailing expression, and a valid statement without. Although it actually gets more complicated than that, because e.g. block is valid as an expression without a trailing expression if there is control flow escaping the current scope

Jackojc replied:
I _think_ I've addressed it by explicitly requiring semicolons on things like declarations and assignments and not requiring statements to end in semicolon anymore but I'll let you read it later and see.

dekrain commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+
+unicode_escape = '\u' hex+
+hex_escape = '\x' hex+
+bin_escape = '\b' ( '0' | '1' )+
```
[DISCUSSION] should this really be a thing?

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+
+# Misc
+expr = match | if | call | cast | type_compare | math
+statement = ( for | while ) | ( ( declaration | expr | assignment ) ';' )
```
Type declarations are listed under the statement category but aren't included here. I don't think they should be included here but they should be under their own "Item" category

Jackojc replied:
For now I've just added them to the statement construct but maybe we can rework and clean up the grammar in the next revision.

dekrain commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+
+call = identifier '(' expr ( ',' expr )* ')' ';'
+
+if = 'if' expr block [ 'else' block ]
```
maybe add support for non-awkward chaining of ~~yandere-dev style code~~ `if else`

Jackojc replied:
Can you elaborate? I think `match` can handle this.

dekrain replied:
I mean something like
```
if ... {
} else if ... {
} else {
}
```
in the first example

noxabellus replied:
> Can you elaborate? I think `match` can handle this.

Match only allows constant expressions

dekrain commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+
+
+# Misc
+expr = match | if | call | cast | type_compare | math
+statement = ( for | while ) | ( ( declaration | expr | assignment ) ';' )
```
[DISCUSSION] this implies that `if` block must end with semicolons, I'd suggest more discussion about this

noxabellus replied:
Yeah I don't think that was the intention. I certainly don't want that. Being explicit about this would require separate grammars for if_expr and if_stmt? Possibly indicates we need block_expr and block_stmt as well, but as I touched on in my comment under the block grammar, its kind of down to semantic differences only that determine if a block is a valid expr

Jackojc replied:
I think I've addressed this but will need to be checked in the new revision.

dekrain reviewed (commented):
Review for ya :slightly_smiling_face:

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+block = '{' statement* [ expr ] '}'
+
+
+program = ( function | sum | enum | multiline_comment )*
```
Missing:
structures
type aliases
global variable declarations
import/export/module? (do we want to wait on this?)

single line comments (presumably comments are allowed everywhere, too)

Jackojc replied:
Yes to everything except module stuff for now. I think single line comments are just going to be handled lexically whereby it just advances until it hits a newline then returns the next valid token. Multiline comments are just a special case to allow nesting.

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+
+call = identifier '(' expr ( ',' expr )* ')' ';'
+
+if = 'if' expr block [ 'else' block ]
```
We need special support for variant matching here
```ebnf
predicate = expr [ 'is' '.' identifier '@' identifier ]
if = 'if' predicate block ( 'else' 'if' predicate block )*  [ 'else' block ]
```

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+
+
+# Misc
+expr = match | if | call | cast | type_compare | math
```
Missing:
`block`
binary ops, unary ops / `math`
array subscript
field access

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+
+# Types
+array_type = '[' integer_literal ( ',' integer_literal )* ']' type
+pointer_type = '^'+ type
```
```ebnf
pointer_type = '^' type
```
Would suffice to convey this grammar, since `pointer_type` is part of `type` ... This doesnt cover the opaque pointer syntax proposed in the discord though. I believe that was just an unadorned `^` with no type following it. So maybe the grammar we want here is
```ebnf
pointer_type = '^' [ type ]
```

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+assignment = identifier '=' ( identifier | literal )
+
+sum_entry = identifier [ '(' type ')' ]
+sum = 'sum' identifier '{' sum_entry ( ',' sum_entry )* [ ',' ] '}'
+
+enum_entry = identifier [ '=' literal ]
+enum = 'enum' '{' enum_entry ( ',' enum_entry )* [ ',' ] '}'
```
I think we should allow defining the backing type for enums and sum discriminators.
e.g.
```
enum MyEnum: u8 { A, B, C }
```
would ensure that `MyEnum` is exactly 1 byte in size

Jackojc replied:
I think we could probably just use the type that is small enough to still cover all of the entries. I don't see anyone really wanting different semantics than that.

noxabellus replied:
I disagree, this was what they thought when they defined the C grammar, and then every compiler had to add hacky attributes to allow this because people prefer to be explicit.

noxabellus replied:
I do think it should be optional though
```ebnf
enum = 'enum' identifier [':' type] '{' enum_entry ( ',' enum_entry )* [ ',' ] '}'
```

Jackojc replied:
Do you think it really makes sense for sum types though?

noxabellus replied:
Absolutely. If we didn't have that we couldn't reliably cross FFI boundaries with sum types

Jackojc replied:
Fair

Unicode-Hater commented on grammar.ebnf:
```diff
@@ -0,0 +1,71 @@
+
+hex_literal = '0x' hex+
+bin_literal = '0b' ( '0' | '1' )+
+integer_literal = ( digit | '_' )+
```
How are we gonna handle inline casting eg:
Rust has something along the line of:
`let value = 42u64;`
While C++ has something like:
`auto value = 42ul`

dekrain replied:
`value := 42;`

noxabellus replied:
Think he was referring to the `u64` / `ul` bit

Unicode-Hater replied:
Ok, but what if i want it to be an explicit 16 bit integer for example, do we have literals for that, or do i need to explicitly cast it

Jackojc replied:
I'm not sure what the best way to handle this is honestly. We could go the route of just requiring you to specify the type if its deduced incorrectly from what you want.

noxabellus replied:
Yeah, `value: u64 = 42` Would be how you did it under the current grammar.
Literals coerce to whichever type is inferred, so most of the time you wouldn't need to state it, outside of declarations

Jackojc committed:
Address requested changes.
- Fixed ack.mlk example not properly using trailing expressions
- Fix initialisation part of for loop
- Fix unicode escape sequences
- Add 'else if' chaining
- Fix syntax highlighting issue with unescaped single quote
- Add utf prefixes to string literals
- Fixed declarations of the form `foo:;`
- Fixed assignment so it allows for assigning to dereferenced pointers and array subscripts
- Add name to enums to support rudimentary namespacing
- Add explicit sizes to both enum and sum structures
- Change type_compare to variant_compare and explicitly require an identifier, not a type
- Fix cast in the case of variants
- Allow calling expressions so we can support array of functions, higher order functions etc.
- Allow arbitrary expressions in match cases
- Allow for multiple, comma seperated match case predicates per block
- Move expr and statement under their respective headings
- Fix for if/match accidentally requiring a terminating semicolon
- Add file-scope level declarations (global variables)
- Add structures, type aliases, globals, goto/label, return, break/continue, math ops, array subscript and field access
- Add variant matching to `if`
- Fix pointer type to use recursive rule

Jackojc commented:
Addressed requested changes in latest commit, please review.

noxabellus commented on grammar.ebnf:
```diff
@@ -10,62 +10,92 @@ comment = ? #.*\n ?
 # Types
-array_type = '[' integer_literal ( ',' integer_literal )* ']' type
-pointer_type = '^'+ type
+array_type = '[' ( integer_literal | identifier ) ( ',' ( integer_literal | identifier ) )* ']' type
```
Missed this before but I think it should be arbitrary exprs for array dimensions
```ebnf
array_type = '[' expr ( ',' expr )* ']' type
```
Any constant expression should be valid here, and the constness is a semantic distinction

Jackojc replied:
I actually updated it to include identifiers which might have made it stand out more haha but I think using expr is probably fair.

noxabellus commented on grammar.ebnf:
```diff
@@ -10,62 +10,92 @@ comment = ? #.*\n ?
+array_type = '[' ( integer_literal | identifier ) ( ',' ( integer_literal | identifier ) )* ']' type
+pointer_type = '^' [ type ]
 fn_type = 'fn' '(' type ( ',' type )* ')' [ type ]
+tuple_type = '(' ( identifier | literal )* ')'
```
Think this is copy pasta from `array_type`? Should be
```ebnf
tuple_type = '(' type ( ',' type )* ')'
```
Where a single element tuple just evaluates to the interior type semantically

noxabellus commented on grammar.ebnf:
```diff
@@ -10,62 +10,92 @@ comment = ? #.*\n ?
+struct_type = 'struct' identifier '{' ( declaration ';' )* '}'
 
-type = array_type | pointer_type | fn_type | identifier
+type = array_type | pointer_type | fn_type | struct_type | identifier
```
Missing `tuple_type`

noxabellus commented on grammar.ebnf:
```diff
@@ -10,62 +10,92 @@ comment = ? #.*\n ?
+struct_type = 'struct' identifier '{' ( declaration ';' )* '}'
 
-type = array_type | pointer_type | fn_type | identifier
+type = array_type | pointer_type | fn_type | struct_type | identifier
```
Personally, I don't think `struct_type` should be here. I don't think anonymous struct types are a desirable feature. This should be with `sum` and `enum`

Jackojc replied:
Yeah good point I completely missed that.

noxabellus commented on grammar.ebnf:
```diff
@@ -10,62 +10,92 @@ comment = ? #.*\n ?
+
+call = expr '(' expr ( ',' expr )* ')'
+
+predicate = expr [ 'is' '.' identifier '@' identifier ]
```
I proposed this grammar in the last review but I realize now it's not quite right. This won't allow multiple `is` bindings, nor will it allow `is` binding to be followed by additional logical operators. Consider these examples:
```
if x is .A @ bind0 and y is .B @ bind1 { .. }
```
```
if x is .A @ bind0 and y == z { .. }
```
```
if x == y and z is .A @ bind0 { .. }
```
I think to resolve this we need to allow `@` bindings in all `is` expressions and leave the validity/usability of the binding as a semantic distinction

noxabellus commented on grammar.ebnf:
```diff
@@ -10,62 +10,92 @@ comment = ? #.*\n ?
 
-call = identifier '(' expr ( ',' expr )* ')' ';'
+array_index = identifier | integer_literal
+array_subscript = identifier '[' array_index ( ',' array_index )* ']'
```
Subscript should allow any expr

noxabellus commented on grammar.ebnf:
```diff
@@ -10,62 +10,92 @@ comment = ? #.*\n ?
+match_predicate = ( ( '.' identifier [ '@' identifier ] ) | expr | '*' )
+match_case =  match_predicate ( ',' match_predicate )* block
 
+bin_op = expr '+' expr
+bin_op = expr '-' expr
+bin_op = expr '*' expr
+bin_op = expr '/' expr
+bin_op = expr '^' expr
+
+unary_op = '-' expr
+unary_op = '+' expr
+unary_op = '^' expr
+unary_op = '@' expr
```
This way of declaring rules is pretty nasty in my opinion

Jackojc replied:
Would you prefer a chain of alternatives using `|`?

noxabellus replied:
What about
```ebnf
bin_op = expr ( '+' | '-' | '*' | '/' | '^' ) expr
```

noxabellus replied:
Though, when we get the full list of ops that might get a bit unwieldy

Jackojc replied:
It's definitely not pretty lol

noxabellus replied:
Another option:
```ebnf
infix = '+' | '-' | '*' | '/' | '^'
      | '==' | '!=' | '<=' | '>=' | '<' | '>'
      | 'and' | 'or'
      | '&' | '|' | '^' | '<<' | '>>'
binop = expr infix expr
```

noxabellus commented on grammar.ebnf:
```diff
@@ -10,62 +10,92 @@ comment = ? #.*\n ?
 match = 'match' expr '{' match_case* '}'
-match_case = ( ( '.' identifier [ '@' identifier ] ) | ( identifier | literal ) | '*' ) block
+match_predicate = ( ( '.' identifier [ '@' identifier ] ) | expr | '*' )
+match_case =  match_predicate ( ',' match_predicate )* block
```
I don't think we want to have multiple @ bindings when using comma separated predicates. The names would all have to be the same to make semantic sense, anyway. I'd prefer to be able to do:
```
match x {
  .A, .B @ bind0 { .. }
  ..
}
```
Not sure how to best state this in grammar from

noxabellus commented on grammar.ebnf:
```diff
@@ -10,62 +10,92 @@ comment = ? #.*\n ?
+bin_op = expr '-' expr
+bin_op = expr '*' expr
+bin_op = expr '/' expr
+bin_op = expr '^' expr
```
Missing logical and bitwise ops

noxabellus replied:
Also missing comparisons

noxabellus commented on grammar.ebnf:
```diff
@@ -10,62 +10,92 @@ comment = ? #.*\n ?
+variant_compare = identifier 'is' '.' identifier
+cast = identifier 'as' ( ( '.' identifier ) | type )
+
+call = expr '(' expr ( ',' expr )* ')'
```
This indicates function calls must have at least one argument

noxabellus commented on grammar.ebnf:
```diff
@@ -10,62 +10,92 @@ comment = ? #.*\n ?
 
-program = ( function | sum | enum | multiline_comment )*
+
+# Misc
```
I think we should rename this section to `# Program` and move the declarations and such here

noxabellus reviewed (changes requested):
Looks like things are progressing well just got a few more details to work out

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,101 @@
+label = 'label' identifier ';'
+goto = 'goto' identifier ';'
+
+statement = ( for | while | alias ) | ( declaration | assignment | expr )
```
This still indicates a `;` is required after `if` and `match`

noxabellus replied:
Actually no this indicates no `;` is required after any expr

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,101 @@
+enum_entry = identifier [ '=' literal ]
+enum = 'enum' identifier [ ':' type ] '{' enum_entry ( ',' enum_entry )* [ ',' ] '}'
+
+function_body = '{' ( statement | ( 'return' expr ';' ) )* '}'
+function = 'fn' identifier '(' declaration ( ',' declaration )* ')' [ type ] function_body
+
+alias = 'type' identifier ':' type ';'
+
+loop_body = '{' ( statement | 'break' | 'continue' )* '}'
```
This is weird. I get that you don't want to allow these control flow keywords in the wrong places but this does weird things like not allow you to return from inside a loop. I would say make `return`, `break`, and `continue` just regular statements, and use regular `block`s instead of these `_body` rules. Yes this leads to some weird cases but they're semantic ones and can't be controlled with grammar in a sensible way

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,101 @@
+
+block = '{' statement* [ expr ] '}'
+
+expr = match | if | call | cast | type_compare | math | block | field_access | array_subscript | literal | operator
```
Missing semantic group expression
```ebnf
sem_group = '(' expr ')'
```

noxabellus replied:
Nevermind, this is covered under `operator`

Jackojc committed:
Fix issues
- Allow arbitrary expressions in array dimensions
- tuple_type now allows arbitrary types
- Add tuple_type to type rule
- Moved struct_type to program rule
- '@' bindings are now valid in all `is` expressions to allow for binding in arbitrary expressions
- Array subscript now accepts an arbitrary expression
- Cleaned up binary and unary op rules
- Add math ops, bitwise ops and comparisons
- Match variant cases now only allow a single binding
- Function calls can now have zero arguments
- Moved sum, enum, struct etc. to a new heading "Program"
- if/match now _really_ don't need `;` after them
- Allow 'return', 'break' and 'continue' in any context, leave it to the semantic analyzer to decide if it's valid

Jackojc commented:
Changes pushed

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,106 @@
+# Program
+declaration = identifier ( ',' identifier )* ( ( ':' type ) | ( ':' [ type ] '=' expr ) ) ';'
+
+function = 'fn' identifier '(' declaration ( ',' declaration )* ')' [ type ] block
```
`declaration` won't work here

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,106 @@
+enum_entry = identifier [ '=' literal ]
+enum = 'enum' identifier [ ':' type ] '{' enum_entry ( ',' enum_entry )* [ ',' ] '}'
+
+struct = 'struct' identifier '{' ( declaration ';' )* '}'
```
`declaration` won't work here

noxabellus commented on grammar.ebnf:
```diff
@@ -0,0 +1,106 @@
+sum_entry = identifier [ '(' type ')' ]
+sum = 'sum' identifier [ ':' type ] '{' sum_entry ( ',' sum_entry )* [ ',' ] '}'
+
+enum_entry = identifier [ '=' literal ]
```
Should be
```ebnf
enum_entry = identifier [ '=' expr ]
```
Any constant expression is valid here

noxabellus reviewed (changes requested):
Nearly there now

Jackojc committed:
More fixes
- Fix function parameters and struct field declaration
- Allow empty structs, enums and sums
- enum_entry now allows arbitrary expressions
- Add true and false keywords

Jackojc committed:
No default args :'v( and moved true/false to expr

Jackojc committed:
Add identifer and nil to expr

noxabellus reviewed (dismissed):
Think this is good to go for the first version's grammar

serbuvlad commented on examples/ack.mlk:
```diff
@@ -0,0 +1,23 @@
+# Ackermann function (http://rosettacode.org/wiki/Ackermann_function)
```
http links should be https?

Jackojc replied:
Unfortunately, rosettacode doesn't seem to use TLS.

serbuvlad commented on grammar.ebnf:
```diff
@@ -0,0 +1,108 @@
+		'^=' | '<<=' | '>>=' | '&=' | '|=' | '~='
+	) expr ';'
+
+for = 'for' identifier ':' [type] '=' expr ',' expr [',' expr] block
```
This is problematic for 2 reasons: The big one is that all example programs that use for would be wrong, because assignments aren't expressions, and do not fit in the third field. On a more minor point, why should we limit the first part of the for to a single variable.

I think the traditional `for non-block-statement; expression; non-block-statement {` is better and cleaner. Would require some grammar reworking in other places.

serbuvlad commented on grammar.ebnf:
```diff
@@ -0,0 +1,108 @@
+
+call = expr '(' [ expr ( ',' expr )* ] ')'
+
+if = 'if' expr block 'else' ( block | if )
```
The else should be optional

Jackojc committed:
Fix for loops and if expressions
- Fix indentation in ack.mlk
- For loops have been simplified and now allow for multi iteration
- If expressions now have an optional else branch

Jackojc committed:
Add infinite loop and loop-until
